returnn/hdf.py

```diff
@@ -346,6 +360,288 @@ def run(self):
+            # avoid having to read the audio file more than once.
+            #
+            # We only pass down the required audio/segment metadata instead of the whole
+            # `i6_core.lib.corpus.Segment` object to avoid excessive pickling overhead,
+            # stalling the worker processes (if we don't).
+            (rec.audio, segments)
+            for rec in c.all_recordings()
+            if len(
+                segments := [
+                    (full_name, segment.start, segment.end)
+                    for segment in rec.segments
+                    if (full_name := segment.fullname()) in all_whitelists
+                ]
+            )
+            > 0
+        )
+        # use imap instead of imap_unordered to have reproducible results, even though this
+        # causes a slight speed penalty
+        for results in pool.imap(apply_func, recs, chunksize=8):
+            for (segment_name, data), (hdf_writer, segments_whitelist) in itertools.product(
+                results, zip(hdf_writers, segment_whitelists)
+            ):
+                if segment_name not in segments_whitelist:
+                    continue
+                _logging.info(f"Writing {segment_name} to {hdf_writer.filename}.")
+                hdf_writer.insert_batch(
+                    inputs=data.reshape(1, -1, 1),
+                    seq_len=[data.shape[0]],
+                    seq_tag=[segment_name],
+                )
+        for hdf_writer in hdf_writers:
+            hdf_writer.close()
+
+    @staticmethod
+    def _process_seq(
+        recording: Tuple[str, Sequence[Tuple[str, float, float]]],
+        *,
+        compress_format: Optional[str],
+        compress_factor: Optional[float],
+        multi_channel_strategy: BlissToPcmHDFJob.BaseStrategy,
+        output_dtype: str,
+        rounding_strategy: BlissToPcmHDFJob.RoundingScheme,
+        round_factor: int,
+        target_sampling_rate: int,
+    ) -> List[Tuple[str, np.ndarray]]:
+        max_tries = 3
+        audio_file, segments = recording
+        for i in range(max_tries):
+            try:
+                return BlissToAudioHDFJob.__process_seq(
+                    audio_file,
+                    segments,
+                    compress_format=compress_format,
+                    compress_factor=compress_factor,
+                    multi_channel_strategy=multi_channel_strategy,
+                    output_dtype=output_dtype,
+                    rounding_strategy=rounding_strategy,
+                    round_factor=round_factor,
+                    target_sampling_rate=target_sampling_rate,
+                )
+            except Exception as exc:
+                if i < max_tries - 1:
+                    _logging.warning(f"Failed to process audio {audio_file}: {exc}")
+                else:
+                    _logging.warning(f"Failed to process audio {audio_file} after {max_tries} attempts. Skipping.")
```

Comment:
This also doesn't seem to be a safe solution, to just skip this? E.g. if `ffmpeg` or so is not correctly installed, it would fail for every seq, and then you get an empty corpus, without error?

Or if the behavior is non-deterministic (I assume so if you try 3 times; why 3 times?), then it means the result of this job here will be non-deterministic, if you just skip it then. This is not good.